core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/encryption/KeyStoreBasedExternalCryptoProvider.java

```diff
@@ -393,14 +402,156 @@ public PrivateKey getPrivateKey(CryptoContext cryptoContext, PrivateKeyInfo priv
+     * @throws CryptoException
+     */
+    @Override
+    public HybridEncryptionOutput hybridEncrypt(HybridEncryptionInput hybridEncryptionInput, String symmetricAlgorithm, String asymmetricAlgorithm,
+                                                String jceSecurityProvider, CryptoContext cryptoContext,
+                                                CertificateInfo certificateInfo) throws CryptoException {
+
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Hybrid encrypt initiated with '%s' symmetric algorithm, '%s' asymmetric " +
+                            "algorithm using '%s' JCE provider related to '%s' crypto context.", symmetricAlgorithm,
+                    asymmetricAlgorithm, jceSecurityProvider, cryptoContext));
+        }
+        String[] resolvedSymmetricAlgorithmSpec = resolveSymmetricAlgorithm(symmetricAlgorithm);
+        int keyLength = Integer.parseInt(resolvedSymmetricAlgorithmSpec[1]) / 8;
+        symmetricAlgorithm = resolvedSymmetricAlgorithmSpec[2];
+        byte[] keyValue = new byte[keyLength];
+        random.nextBytes(keyValue);
+        SecretKeySpec symmetricKey = new SecretKeySpec(keyValue, resolvedSymmetricAlgorithmSpec[0]);
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("A secret key of %s type was successfully generated for hybrid encryption.",
+                    resolvedSymmetricAlgorithmSpec[0]));
+        }
+        try {
+            Cipher cipher;
+            if (StringUtils.isBlank(jceSecurityProvider)) {
+                cipher = Cipher.getInstance(symmetricAlgorithm);
+            } else {
+                cipher = Cipher.getInstance(symmetricAlgorithm, jceSecurityProvider);
+            }
+            AlgorithmParameterSpec algoParams = AlgorithmParameterResolver
+                    .resolveSymmetricAlgorithmParameters(symmetricAlgorithm);
+            cipher.init(Cipher.ENCRYPT_MODE, symmetricKey, algoParams);
+
+            if (hybridEncryptionInput.getAuthData() != null) {
+                cipher.updateAAD(hybridEncryptionInput.getAuthData());
+            }
+            byte[] encryptedData = cipher.doFinal(hybridEncryptionInput.getPlainData());
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Plain data was successfully encrypted with %s symmetric algorithm and %s " +
+                        "secret key.", symmetricAlgorithm, resolvedSymmetricAlgorithmSpec[0]));
+            }
+            byte[] encryptedKey = encrypt(keyValue, asymmetricAlgorithm, jceSecurityProvider, cryptoContext,
+                    certificateInfo);
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Secret key value was successfully encrypted with %s asymmetric algorithm " +
+                        "with public certificate : %s", asymmetricAlgorithm, certificateInfo));
+            }
+            if (hybridEncryptionInput.getAuthData() != null) {
```

Comment:
This logic is not readable. Move the logic to a method which has a meaningful name for the logic.

Reply:
Separated the symmetric encryption logic into another private method.